Source/Core/Core/IOS/Network/IP/Top.cpp

```diff
@@ -162,70 +167,223 @@ static s32 MapWiiSockOptNameToNative(u32 optname)
+    const std::span<MIB_IPFORWARDROW> table(forward_table->table, forward_table->dwNumEntries);
+    for (const auto& entry : table)
     {
-      if (forward_table->table[i].dwForwardDest == 0)
+      routing_table.emplace_back(entry.dwForwardIfIndex,
+                                 std::bit_cast<in_addr>(entry.dwForwardDest),
+                                 std::bit_cast<in_addr>(entry.dwForwardMask),
+                                 std::bit_cast<in_addr>(entry.dwForwardNextHop));
+    }
+
+    if (result == NO_ERROR)
+      break;
+
+    result = GetIpForwardTable(forward_table.get(), &forward_table_size, FALSE);
+  }
+
+#elif defined(__linux__)
+  constexpr int BUFF_SIZE = 8192;
+  unsigned int msg_seq = 0;
+  const int sock = socket(PF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
+  if (sock < 0)
+  {
+    ERROR_LOG_FMT(IOS_NET, "Failed to open netlink socket, error {}", Common::StrNetworkError());
+    return {};
+  }
+
+  Common::ScopeGuard socket_guard{[sock] { close(sock); }};
+  std::array<char, BUFF_SIZE> msg_buffer{};
+  auto nl_msg = reinterpret_cast<nlmsghdr*>(msg_buffer.data());
+  auto rt_msg = reinterpret_cast<rtmsg*>(NLMSG_DATA(nl_msg));
+  const unsigned int pid = getpid();
+
+  // prepare command/netlink packet
+  nl_msg->nlmsg_len = NLMSG_LENGTH(sizeof(rtmsg));  // Length of message.
+  nl_msg->nlmsg_type = RTM_GETROUTE;                // Get the routes from kernel routing table .
+
+  nl_msg->nlmsg_flags = NLM_F_DUMP | NLM_F_REQUEST;  // The message is a request for dump.
+  nl_msg->nlmsg_seq = msg_seq++;                     // Sequence of the message packet.
+  nl_msg->nlmsg_pid = pid;                           // PID of process sending the request.
+
+  // ship it
+  if (send(sock, nl_msg, nl_msg->nlmsg_len, 0) < 0)
+  {
+    ERROR_LOG_FMT(IOS_NET, "Failed to send netlink request ({})", Common::StrNetworkError());
+    return {};
+  }
+
+  // read response
+  int msg_len = 0;
+  msg_buffer.fill(0);
+
+  do
+  {
+    auto buf_ptr = msg_buffer.data() + msg_len;
+    const int read_len = recv(sock, buf_ptr, BUFF_SIZE - msg_len, 0);
+    if (read_len < 0)
+    {
+      ERROR_LOG_FMT(IOS_NET, "Failed to receive netlink response ({})", Common::StrNetworkError());
+      return {};
+    }
+
+    nl_msg = reinterpret_cast<nlmsghdr*>(buf_ptr);
+    if (NLMSG_OK(nl_msg, read_len) == 0)
+    {
+      ERROR_LOG_FMT(IOS_NET, "Received netlink error response ({})", NLMSG_OK(nl_msg, read_len));
+      return {};
+    }
+
+    if (nl_msg->nlmsg_type == NLMSG_ERROR)
+    {
+      auto err = reinterpret_cast<nlmsgerr*>(NLMSG_DATA(nl_msg));
+      if (nl_msg->nlmsg_len < NLMSG_LENGTH(sizeof(nlmsgerr)))
       {
-        ifIndex = forward_table->table[i].dwForwardIfIndex;
-        break;
+        ERROR_LOG_FMT(IOS_NET, "Netlink message was truncated");
+      }
+      else
+      {
+        ERROR_LOG_FMT(IOS_NET, "Received netlink error ({})",
+                      Common::DecodeNetworkError(err->error));
```

Comment:
The original error value should also be passed with the message. The error also need to be negated to be properly decoded:
```C++
ERROR_LOG_FMT(IOS_NET, "Received netlink error {}: {}", err->error, Common::DecodeNetworkError(-err->error));
```

I managed to confirm that by providing an invalid `nlmsg_type` which produces:
> E[IOS_NET]: Received netlink error -95: Operation not supported